### DISCUSSION TITLE
Schema or Grammar for Policy Definition and result files

### DISCUSSION THREAD

**Discussion by jgsuess** [oracle/macaron#218] [Q&A]
I was looking to create an example of a model-driven tool-set and was wondering if there is a schema or grammar for either the policy definition that the verifier uses, or the analysis result. Having a schema or grammar greatly simplifies the construction. If the docs will contain this information I will build the docs and have a look, but I wanted to check first to see if this is worthwhile.

**Comment by behnazh-w** [comment]
The two main policy languages that we currently support are [Datalog](https://github.com/oracle-samples/macaron/blob/d6622ee4f53b0d4d4d90a9c3e81306015b963b40/src/macaron/policy_engine/policies.md) and [CUE](https://cuelang.org/). We also have a [simple policy](https://github.com/oracle-samples/macaron/blob/d6622ee4f53b0d4d4d90a9c3e81306015b963b40/src/macaron/policy_engine/policy_schema.yaml) but it's less powerful than the others. For Datalog we use Souffle as the solver. In our current design check results are stored in a database and passed to Souffle to enforce a policy. [Here](https://github.com/oracle-samples/macaron/blob/d6622ee4f53b0d4d4d90a9c3e81306015b963b40/docs/source/assets/er-diagram.svg) is an ERD for our current checks but we are changing our data model, so it's best to wait for the new version.

We are actively improving our documentations and will add various tutorials to help users understand different aspects of Macaron. It's great to hear your feedback.

I'm curious about your concrete usecase. Can you please provide an example?

**Comment by jgsuess** [comment]
I am looking for a good and salient example immediately applicable in the context of SE research on how to apply [the Epsilon](https://www.eclipse.org/epsilon/) systems in combination with [MDE for DevOps](https://www.researchgate.net/publication/356491511_Using_DevOps_Toolchains_in_Agile_Model-Driven_Engineering). I wanted to construct an example that shows how the tools can be used to provide a well-guided configuration for first time users of advanced bespoke software tools. Macaron meets both of those criteria.
I see a few use-cases:
1. Execution configuration and reporting

* Defining execution parameters for runs as an input model (tree-based, option-based), including advice and validation
* Producing execution command lines of such parameter sets
* Parsing results
* Visualising and querying evaluation results
* Producing a salient overall run report that can be adapted with small changes
(This was the docker control use case.)

2. Policy definition for non-language experts
* Defining reasonable quick-start policies using an option-based or tree based system
* Formatted display of such policies for easy consumption and in embedding in automated execution
* Visualising and querying evaluation results relative to the policy input

The ERD is closest to what I am looking at. In fact, the Epsilon/EMF toolset is *exactly* designed for clearly defining such data structures.

Sorry for the long answer. I am really not trying to take away time or distract.
